File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
  int check(TreeNode root,int[] maxi){
        if(root==null) return 0;

        int left=Math.max(0,check(root.left,maxi));
        int right=Math.max(0,check(root.right,maxi));

        maxi[0]=Math.max(maxi[0],root.val+left+right);

        return root.val+Math.max(left,right);

    }
    public int maxPathSum(TreeNode root) {
        int[] maxi=new int[1];
        maxi[0]=-(int)(1e9);
        int a=check(root,maxi);
        //this function is giving tha max sum path either left or right bas not the ans
        return maxi[0];
    }
 */
class Solution {
    int check(TreeNode root,int[] ans){
        if(root==null) return 0;

        int ls=0,rs=0;
        ls=check(root.left,ans);
        rs=check(root.right,ans);
        
        int a=0;
        if(ls<0 && rs<0) a=0;
        else if(ls<=0 || rs<=0) a=Math.max(ls,rs);
        else a=ls+rs;

        ans[0]=Math.max(ans[0],a+root.val);

        return root.val+Math.max(Math.max(0,ls),Math.max(rs,0));
    }
    public int maxPathSum(TreeNode root) {
        int[] ans=new int[1];
        ans[0]=-(int)1e9;
        int k=check(root,ans);
        return ans[0];
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        def check(node, ans):
            if node == None:
                return 0
            
            lsum = max(check(node.left, ans), 0)
            rsum = max(check(node.right, ans), 0)
            
            k = node.val + rsum + lsum
            ans[0] = max(ans[0], k)
            return node.val + max(lsum, rsum)
        
        ans = [-(10 ** 10)]
        check(root, ans)
        return ans[0]
        